### ISSUE TITLE
GPU backend (WebGPU/OpenCL via ml_drift) silently omits trailing generated content after a tag-like boundary; CPU backend produces correct output on identical model

### ISSUE THREAD

**Issue by bosejapan23-06** [google-ai-edge/LiteRT-LM#3083]
## Summary
On the GPU backend, a fine-tuned Qwen2.5-1.5B model trained to produce
structured output in the form `<MOOD>state</MOOD> <TOOL>toolName(args)</TOOL>`
frequently stops generation immediately after the first tag
(`</MOOD>`), omitting the second tag (`<TOOL>...</TOOL>`) that the
identical .litertlm file reliably produces on the CPU backend for the same
input. Across a 60-case evaluation suite, GPU tool-tag correctness is
35/60 vs. 56/60 on CPU -- same model file, same prompts.

## Environment
- litert-lm: 0.14.0 (confirmed current latest release)
- litert-torch-nightly: 0.10.0.dev20260725
- Backend tested: GPU (WebGPU/OpenCL via ml_drift), compared against CPU
- Model: fine-tuned Qwen2.5-1.5B-Instruct, LoRA-merged, exported via
  litert-torch export_hf with the StableHLOCompositeBuilder neutralization
  patch (see issue https://github.com/google-ai-edge/litert-torch/issues/1136) applied, quantization_recipe=dynamic_wi8_afp32,
  prefill_lengths=[32,128,512,1280]

## Steps to Reproduce
1. Export a fine-tuned Qwen2.5-1.5B model with the above configuration.
2. Run identical prompts through both backends:
   litert-lm run model.litertlm --backend=cpu --prompt="<real prompt>"
   litert-lm run model.litertlm --backend=gpu --prompt="<real prompt>"
3. Compare outputs across many cases where the expected output includes
   two sequential tags.

## Expected
Identical (or functionally equivalent) output between CPU and GPU backends
for the same deterministic generation request.

## Actual
GPU frequently terminates generation right after the first tag closes,
even though the model's own reasoning (the preceding sentence) clearly
indicates it intended to continue. CPU produces the complete, correct
output for the identical input every time in our test suite.

Example (identical prompt, both backends):
  CPU: "Of course, I'm warming it up for you. <MOOD>happy</MOOD> <TOOL>increaseTemperature(all)</TOOL>"
  GPU: "Of course, I'm warming it up for you. <MOOD>happy</MOOD>"

## What we've ruled out
- Random sampling: confirmed deterministic via repeated identical runs.
- Generation-length limit: --max-num-tokens is documented as the KV-cache
  size, not an output-length limiter; confirmed no change when explicitly
  maximized.
- Speculative decoding: --enable-speculative-decoding=false produces
  byte-identical output to the default; ruled out.
- Quantization recipe: tested weight_only_wi8_afp32 (different, unrelated
  crash), dynamic_wi8b32_afp32 and dynamic_wi8b64_afp32 (both silently
  apply zero quantization in our installed version -- Quantization Ratio:
  1.00 in the export log -- so not a valid data point), and
  dynamic_wi8c_afp32 (per-channel granularity -- applies real, correct
  quantization at the expected size, but exhibits the identical bug).
  The one configuration that DID produce correct GPU output used no
  quantization at all (full fp32/bf16, ~6.6GB, not deployable at that
  size), which is consistent with -- but not proof of -- a
  quantization-precision-related root cause specifically in GPU kernel
  execution.
- Prefill bucket structure: tested a tighter-fitting bucket set: no change.

## Hypothesis
Given content and mood are consistently correct and only this one specific
decode-position decision differs, we suspect the GPU delegate's floating-
point computation (different accumulation order/kernel implementation vs.
CPU) produces a logit distribution that's sufficiently different from CPU's
at this one close-call decision point (continue vs. end-of-sequence) to
flip the argmax outcome, even though both backends agree confidently
everywhere else in the sequence.

## Question
Is this a known characteristic of the ml_drift GPU delegate's numerical
precision vs. the CPU/XNNPack path? Is there a recommended way to bias
generation to reduce sensitivity to this kind of near-tie at tag/structural
boundaries, or a delegate-level fix planned?

Happy to provide the exact model export command, full evaluation logs, or
run additional targeted reproduction steps if useful.

**Comment by bosejapan23-06** [comment]
Requesting maintainer to add "Labels: bug" to this issue.

**Comment by bosejapan23-06** [comment]
Hi @nireekshak ,
Thank you for your support. Could you possibly provide an update ?

**Comment by john-rocky** [comment]
Following up from google-ai-edge/litert-torch#1136. I don't have a confirmed root cause for this one, but the shape of it — GPU silently stopping at a structural boundary while CPU stays correct, and the only fully-correct GPU config being the unquantized export — matches a family of backend-divergence behaviors I've been measuring elsewhere. So: the triage I would run, plus an offer.

Probably the highest-value single step: litert-lm v0.16.0 was released yesterday, and the commit history between v0.14.0 and v0.16.0 contains several GPU-path fixes in exactly this area — e.g. "Fix subtle sync issue in input handling on WebGPU" and "Set param tensor non-external as WebGPU has some sync issue between 2 inferences when input tensors are external". Re-running your 60-case suite on 0.16.0 on both backends is the cheapest way to learn whether this is already fixed.

Details, if useful:

- What the GPU actually emits at the stop: if you can log token ids, the key question is whether the GPU produces an end-of-turn/eos token right after `</MOOD>`, or just stops streaming. A single low-margin eos flip would point at small GPU numeric differences amplified by the quantized weights — consistent with your fp32-works observation.
- For one failing prompt, vary prefill_lengths / prompt length: if the stop position tracks a prefill-bucket boundary rather than the tag, it's a prefill/decode handoff issue rather than a logits flip — a different bug.
- Since dynamic_wi8 is the common factor in every failing config, a weight-only int8 data point would be valuable — you mentioned an unrelated crash there; if you can share that log I'd take a look (in other graphs I've measured, dynamic-range int8 degraded where weight-only int8 was near-lossless at the same size).
- If you can share the .litertlm (or an export command from a public finetune of the same shape), I'm happy to run it through the backend-parity gate I use before releases (Mac GPU + Pixel 8a) and report the token-level divergence positions back here.

No urgency on my side — and if the 0.16.0 rerun already fixes it, even better.
